designs/vpc_load_balancing.md

```diff
@@ -0,0 +1,84 @@
+be part of the same VPC. Since AWS has a [quota on the maximum number of VPC Endpoints per VPC](https://docs.aws.amazon.com/vpc/latest/userguide/amazon-vpc-limits.html#vpc-limits-endpoints),
+which starts at 50 and is generally easy to raise to 200, when AVO is heavily used at scale it can run into this limit.
+
+```go
+// Vpc represents the configuration for the AWS VPC to create the VPC Endpoint in
+type Vpc struct {
+    // +kubebuilder:validation:Optional
+
+    // AutoDiscoverSubnets will instruct the controller to use the subnets associated with this ROSA cluster if true.
+    AutoDiscoverSubnets bool `json:"autoDiscoverSubnets,omitempty"`
+
+    // SubnetIds is a list of subnet ids to associate with the VPC Endpoint, which must all be in the same VPC.
+    // If more than one is specified, each subnet must be in a different Availability Zone.
+    // Ref: https://docs.aws.amazon.com/vpc/latest/privatelink/create-interface-endpoint.html
+    SubnetIds []string `json:"subnetIds,omitempty"`
+}
+```
+
+## Motivation
+
+This change is important because:
+
+- It will prevent the need of a higher level orchestrator to stamp out `VpcEndpoint` CRs load balanced across a set of VPCs
+
+## Relevant Stories
+
+- [OSD-13906](https://issues.redhat.com/browse/OSD-13906) Creating VPC Endpoints in the OSD Transit Account
+- [OSD-15012](https://issues.redhat.com/browse/OSD-15012) Load Balancing VPC Endpoints across VPCs
+
+## Goals
+
+- Add an optional set of VPC IDs specified in `.spec.vpc.ids`
+- Pick out private subnet IDs from the provided list of VPCs with the aid of `kubernetes.io/role/internal-elb`
+  - Filter out relevant subnets that are in Availability Zones that the VPC Endpoint Service supports
+- Bin pack VPC Endpoints into relevant VPCs and providing metrics to help inform when new VPCs or additional quota is needed
+
+## Non-Goals/Future Work
+
+- Have AVO dynamically create and/or manage VPCs directly
+
+## Proposal
+
+New fields in `.spec.vpc`:
+
+```go
+// Vpc represents the configuration for the AWS VPC to create the VPC Endpoint in
+type Vpc struct {
+    // +kubebuilder:validation:Optional
+
+    // AutoDiscoverSubnets will instruct the controller to use the subnets associated with this ROSA cluster if true.
+    AutoDiscoverSubnets bool `json:"autoDiscoverSubnets,omitempty"`
+
+    // SubnetIds is a list of subnet ids to associate with the VPC Endpoint, which must all be in the same VPC.
+    // If more than one is specified, each subnet must be in a different Availability Zone.
+    // Ref: https://docs.aws.amazon.com/vpc/latest/privatelink/create-interface-endpoint.html
+    SubnetIds []string `json:"subnetIds,omitempty"`
+	
+    // Ids is a list of VPC ids that aws-vpce-operator can choose from to load balance. All provided VPCs must be in the
+    // same region as the specified VPC Endpoint Service (.spec.serviceName).
+    Ids []string `json:"ids,omitempty"`
```

Comment:
What's a good way of specifying quota per VPC? Should AVO just have permissions to read the quota itself?

Reply:
I think quota discovery should be out of scope for AVO. Monitoring of the transit accounts for quota exhaustion already exists. Or is this to determine if a VPC is at capacity?

Reply:
Since we have existing monitoring in place for quota exhaustion, dropping this concern around reading the quota